packages/devtools/src/components/CardDesigner/CardDesignerEditor.tsx

```diff
@@ -0,0 +1,196 @@
+
+import { useCardDesignerEditorClasses } from './CardDesignerEditor.styles';
+import type { SelectTabData, SelectTabEvent, TabValue } from '@fluentui/react-components';
+
+export interface CardDesignerEditorProps {
+  readonly value?: Card;
+  readonly typescript?: string;
+  readonly onChange?: (value: Card) => void;
+}
+
+export default function CardDesignerEditor({
+  value,
+  typescript,
+  onChange,
+}: CardDesignerEditorProps) {
+  const classes = useCardDesignerEditorClasses();
+  const [selectedValue, setSelectedValue] = useState<TabValue>('json');
+
+  const onTabSelect = (_event: SelectTabEvent, data: SelectTabData) => {
+    setSelectedValue(data.value);
+  };
+
+  return (
+    <div className={classes.tabGroup}>
+      <TabList selectedValue={selectedValue} onTabSelect={onTabSelect}>
+        <Tab id="json" value="json">
+          Json
+        </Tab>
+        <Tab id="typescript" value="typescript">
+          Typescript
+        </Tab>
+      </TabList>
+      <div className={classes.tabPanels}>
+        {selectedValue === 'json' && (
+          <div role="tabpanel" aria-labelledby="json" className={classes.tabPanel}>
+            <CardDesignerJsonEditor value={value} onChange={onChange} />
+          </div>
+        )}
+        {selectedValue === 'typescript' && (
+          <div role="tabpanel" aria-labelledby="typescript" className={classes.tabPanel}>
+            <CardDesignerTypescriptEditor value={typescript} />
+          </div>
+        )}
+      </div>
+    </div>
+  );
+}
+
+export interface CardDesignerJsonEditorProps {
+  readonly value?: Card;
+  readonly onChange?: (value: Card) => void;
+}
+
+export function CardDesignerJsonEditor({ value, onChange }: CardDesignerJsonEditorProps) {
+  const ref = useRef<HTMLDivElement>(null);
+  const viewRef = useRef<EditorView | null>(null);
+  const initializedRef = useRef(false);
+  const classes = useCardDesignerEditorClasses();
+  const [isUpdating, setIsUpdating] = useState(false);
+
+  // Initialize the editor once
+  useEffect(() => {
+    if (!ref.current || initializedRef.current || viewRef.current) return;
+
+    initializedRef.current = true;
+
+    const view = new EditorView({
+      parent: ref.current,
+      state: EditorState.create({
+        doc: value ? JSON.stringify(value, null, 2) : undefined,
+        extensions: [
+          basicSetup,
+          atomone,
+          json(),
+          EditorView.updateListener.of((update: ViewUpdate) => {
+            if (!update.docChanged || !onChange || isUpdating) return;
+
+            try {
+              const card = JSON.parse(update.state.doc.toString());
+              onChange(card);
+            } catch (err) {
+              console.error(err);
+            }
+          }),
+        ],
+      }),
+    });
+
+    viewRef.current = view;
+
+    return () => {
+      view.destroy();
+      viewRef.current = null;
+      initializedRef.current = false;
+    };
+  }, [ref.current]);
```

Comment:
FWIW `ref.current` as a dependency will always only work as a side effect, meaning if re-rendering is happening due to some other state change.

Reply:
Logged for fast-follow https://github.com/microsoft/spark.js/issues/18